### DISCUSSION TITLE
Order of columns in INSERT INTO appears random for createMany

### DISCUSSION THREAD

**Discussion by Jens-Rydholm-Humly** [prisma/prisma#22301] [Q&A] [answered]
### Question

I am monitoring the SQL generated by Prisma (using `pg_stat_statements` in PostgreSQL), and I've encountered a behaviour that makes this harder.

When using Prisma's `createMany` operation, it appears as if the column order of the executed INSERT INTO statements is essentially random. Each execution gets a different column order, making them separate queries when they are actually the same piece of code. Example (anonymised, and with only one set of values each):
* INSERT INTO "my_table" ("foo", "bar", "fum") VALUES ($1,$2,$3)
* INSERT INTO "my_table" ("foo", "fum", "bar") VALUES ($1,$2,$3)
* ...

From what I've seen so far, this is not the case when using the `upsert` operation of Prisma. In those cases, the column order appears to be the same as that of the Prisma schema file, making the SQL consistent.

Is this difference intentional? It makes monitoring harder (at least for my chosen approach), and it seems like it might affect e.g. DBMS query plan caching as well. Any suggestions for how to proceed? I would prefer not using separate `upsert`/`create` calls in my code, as a bulk insert should be more efficient.

I am using Prisma 5.5.2 right now.

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: Linux
- Database: PostgreSQL
- Node.js version: v20.3.1

prisma                  : 5.5.2
@prisma/client          : 5.5.2
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine aebc046ce8b88ebbcb45efe31cbe7d06fd6abc0a (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Schema Engine           : schema-engine-cli aebc046ce8b88ebbcb45efe31cbe7d06fd6abc0a (at node_modules/@prisma/engines/schema-engine-darwin-arm64)
Schema Wasm             : @prisma/prisma-schema-wasm 5.5.1-1.aebc046ce8b88ebbcb45efe31cbe7d06fd6abc0a
Default Engines Hash    : aebc046ce8b88ebbcb45efe31cbe7d06fd6abc0a
Studio                  : 0.494.0

**Comment by janpio** [comment]
Not intentional, please open a bug report if one does not exist. Thanks.

**Reply by Jens-Rydholm-Humly** [reply]
Thanks, will do!

**Reply by Jens-Rydholm-Humly** [reply, marked as answer]
Bug report filed here: https://github.com/prisma/prisma/issues/22307
I created a public GitHub repo reproducing the problem.
